Docs: Update OIDC SSO Guide (#29087)
gravitational/teleport: docs/pages/access-controls/sso/oidc.mdx

---
title: OAuth2 and OIDC authentication

> description: How to configure Teleport access with OAuth2 or OIDC (OpenID connect)

Reviewer: Isn't it more common to say "Open ID Connect (OIDC)"

Author: ```suggestion
description: How to configure Teleport access with OAuth2 or OpenID connect (OIDC)
```

---

This guide will explain how to configure an SSO provider using [OpenID
Connect](http://openid.net/connect/) (also known as OIDC) to issue Teleport

> credentials to a specific groups of users. When used in combination with role
> based access control (RBAC) it allows Teleport administrators to define
> policies like:

Reviewer 2: ```suggestion
credentials to specific groups of users. When used in combination with role-based 
access control (RBAC), OIDC allows Teleport administrators to define
policies like:
```

- Only members of the "DBA" group can connect to PostgreSQL databases.
- Developers must never SSH into production servers.

## Prerequisites

- Admin access to the SSO/IdP being integrated with users assigned to groups/roles.

> - Teleport role with access to maintaining `oidc` resources. This is available
>   in the default `editor` role.

Reviewer 2: ```suggestion
- Teleport role with permission to maintain `oidc` resources. This permission is available
  in the default `editor` role.
```
The "access to maintaining" sounds a little odd to my ear. I would guess it is more like a role granted x permissions.

- [Google Identity Platform](https://developers.google.com/identity/protocols/OpenIDConnect)
- [Keycloak Client Registration](https://www.keycloak.org/docs/latest/securing_apps/index.html#\_client_registration)

> <Admonition type="tip">

Reviewer: Should we add a title "Google Workspace" to this admonition?

As a reader I often find titles helpful because it quickly let's me know whether I can skip over the admonition or whether it applies to me and I should fully understand it.

Author: ```suggestion
<Admonition type="tip" title="Google Workspace">
```

For Google Workspace please follow our dedicated [Guide](google-workspace.mdx)
</Admonition>

Save the relevant information from your identity provider. To make following
this guide easier, you can add the values here and they will be included in the
example commands below:

- Client ID: <Var name="<CLIENT-ID>" description="The Client ID for your Teleport application, provided by your IdP"/>
- Client secret: <Var name="<CLIENT-SECRET>" description="The Client secret for your Teleport application, provided by your IdP"/>

## OIDC Redirect URL

OIDC relies on HTTP redirects to return control back to Teleport after
authentication is complete. The redirect URL must be selected by a Teleport
administrator in advance.

The redirect URL for OIDC authentication in Teleport is <nobr><Var
name="mytenant.teleport.sh" description="Your Teleport Cloud tenant or Proxy
Service address"/>`/v1/webapi/oidc/callback`</nobr>. Replace `mytenant.teleport.sh`
with your Teleport Cloud tenant or Proxy Service address.

## OIDC connector configuration

The next step is to add an OIDC connector to Teleport. The connectors are

> created, test, and added / removed via `tctl` [resource commands](../../reference/resources.mdx).

Reviewer: Or via the web UI (not sure if that's worth mentioning here)

To create a new connector, use `tctl sso configure`. Our example creates a
connector resource file in YAML format named `oidc-connector.yaml`:

```code
$ tctl sso configure oidc --name <CONNECTOR-NAME> \
  --issuer-url <PATH-TO-PROVIDER> \

>   --id <CLIENT-ID> --secret <CLIENT-SECRET> \

Reviewer: This has the side effect of leaving the client secret in the user's shell history, which isn't great. I wonder if we should suggest an alternative.

Author: Is there an alternative that still creates a complete connector resource?

How about a warning to consider scrubbing this command from shell history? The `--secret` key is used in several other pages where `tctl sso` is used, so we could make it a partial file.

Reviewer: I don't think so, but we could have the user put the secret in a file and use `--secret $(cat client-secret.txt` or something.

Author: Since the use of `--secret` is now in several docs, I'll make a new issue that once this PR (#29087) is merged, we should update all the SSO guides to use this method.

> I don't think so, but we could have the user put the secret in a file and use `--secret $(cat client-secret.txt)` or something.

  --claims-to-roles <CLAIM-KEY>,<CLAIM-VALUE>,access \
  --claims-to-roles <CLAIM-KEY>,<CLAIM-VALUE>,editor > oidc-connector.yaml
```

- `--name`: Usually the name of the IdP, this is how the connector will be
  identified in Teleport.
- `--issuer-url`: This is the base path to the IdP's OIDC configuration endpoint,
  excluding `.well-known/openid-configuration`. If, for example, the endpoint
  is `https://example.com/.well-known/openid-configuration`, you would use

>   `example.com`.

Reviewer: Wouldn't it be https://example.com then? (example.com is not a URL, nor does it match the keycloak example below)

Author: ```suggestion
  `https://example.com`.
```

- [Google Identity Platform](https://developers.google.com/identity/protocols/OpenIDConnect)
- [Keycloak Client Registration](https://www.keycloak.org/docs/latest/securing_apps/index.html#\_client_registration)

<Admonition type="tip" title="Google Workspace">

> For Google Workspace please follow our dedicated [Guide](google-workspace.mdx)

Reviewer 2: ```suggestion
For Google Workspace, see [Teleport Authentication with Google Workspace](google-workspace.mdx)
```

</Admonition>

Save the relevant information from your identity provider. To make following
this guide easier, you can add the values here and they will be included in the
example commands below:

- Client ID: <Var name="<CLIENT-ID>" description="The Client ID for your Teleport application, provided by your IdP"/>
- Client secret: <Var name="<CLIENT-SECRET>" description="The Client secret for your Teleport application, provided by your IdP"/>

## OIDC Redirect URL

OIDC relies on HTTP redirects to return control back to Teleport after
authentication is complete. The redirect URL must be selected by a Teleport
administrator in advance.

The redirect URL for OIDC authentication in Teleport is <nobr><Var
name="mytenant.teleport.sh" description="Your Teleport Cloud tenant or Proxy
Service address"/>`/v1/webapi/oidc/callback`</nobr>. Replace `mytenant.teleport.sh`
with your Teleport Cloud tenant or Proxy Service address.

## OIDC connector configuration

The next step is to add an OIDC connector to Teleport. The connectors are

> created, test, and added / removed via `tctl` [resource commands](../../reference/resources.mdx)
> or the Web UI.

Reviewer 2: ```suggestion
created, test, and added or removed using `tctl` [resource commands](../../reference/resources.mdx)
or the Teleport Web UI.
```

> To create a new connector, use `tctl sso configure`. Our example creates a

Reviewer 2: ```suggestion
To create a new connector, use `tctl sso configure`. The following example creates a
```

connector resource file in YAML format named `oidc-connector.yaml`:

```code
$ tctl sso configure oidc --name <CONNECTOR-NAME> \
  --issuer-url <PATH-TO-PROVIDER> \
  --id <CLIENT-ID> --secret <CLIENT-SECRET> \
  --claims-to-roles <CLAIM-KEY>,<CLAIM-VALUE>,access \
  --claims-to-roles <CLAIM-KEY>,<CLAIM-VALUE>,editor > oidc-connector.yaml
```

- `--name`: Usually the name of the IdP, this is how the connector will be
  identified in Teleport.
- `--issuer-url`: This is the base path to the IdP's OIDC configuration endpoint,
  excluding `.well-known/openid-configuration`. If, for example, the endpoint
  is `https://example.com/.well-known/openid-configuration`, you would use
  `https://example.com`.
- `--id`: The client ID as defined in the IdP. Depending on your identity

>   provider this may be something you can define (i.e. `teleport`), or may be an

Reviewer 2: ```suggestion
  provider this may be something you can define (for example, `teleport-oidc`), or may be an
```
It seems like if it's something I can define, I wouldn't use the generic/non-descriptive name `teleport` but maybe I'm missing something.

Author: In this case if you were defining this value, it would be within the context of the OIDC provider, so the name of the client ID would be relative to that context. In that context, the `-oidc` would be redundant since you're making it in the OIDC provider, and only specifying the name of the application consuming it.

```suggestion
  provider this may be something you can define (for example, `teleport`), or may be an
```

  assigned string.
- `--secret`: The client token/secret provided by the IdP to authorize this client.
- `--claims-to-roles`: A mapping of OIDC claims/values to be associated with
  Teleport roles.

For more information on these and all available flags, see the [tctl sso configure
oidc](docs/reference/cli.mdx#tctl-sso-configure-oidc) section of the Teleport CLI
Reference page.

The file created should look like the example below. This connector requests
the scope `<CLAIM-KEY>` from the identity provider, then maps the value to
either the `access` or the `editor` role depending on the value returned for
that key within the claims:

```yaml
(!examples/resources/oidc-connector.yaml!)
```

<Details title="Practical Example: Keycloak">
The following example was generated using Keycloak as the identity provider.
Keycloak is being served at `keycloak.example.com`, and the Teleport Proxy
Service is listening at `teleport.example.com`. In Keycloak, the client is

> named `teleport`, and under the `teleport-dedicated` client scope we've added
> the "Group Membership" mapper:

Reviewer 2: ```suggestion
named `teleport`. Under the `teleport-dedicated` client scope, we've added
the "Group Membership" mapper:
```
I don't really understand this-probably just missing context—but I don't know where ``teleport-dedicated` client scope` is or where `the "Group Membership" mapper` is.

Author: Yes, all the context for this lives in the Keycloak UI. I added it because Keycloak's UI is unintuitive IMO and this instruction will help them navigate where they need to do. If Keycloak were more popular a dedicated guide might be in order.

```yaml
kind: oidc
metadata:
  name: keycloak
spec:
  claims_to_roles:
  - claim: groups
    roles:
    - access
    value: /users
  - claim: groups
    roles:
    - editor
    value: /admins
  client_id: teleport
  client_secret: abc123...
  issuer_url: https://keycloak.example.com/realms/master
  redirect_url: https://teleport.example.com:443/v1/webapi/oidc/callback
version: v3

```

</Details>

Before applying the connector to your cluster, you can test that it's configured
correctly:

```code
$ cat oidc-connector | tctl sso test
```

This should open up your web browser and attempt to log you in to the Teleport
cluster through your IdP. If it fails, review the output of this command for
troubleshooting details.

<Admonition type="tip">
The "[OIDC] Claims" section of the CLI output provides all the details of your
user provided by the IdP. This is a good starting point while troubleshooting
errors like `Failed to calculate user attributes.`
</Admonition>

> One your tests are successful, create the connector:

Reviewer 2: ```suggestion
After your tests are successful, create the connector:
```

  # redirect_timeout: 90s
```

### Optional: Disable email verification

> By default, Teleport will validate the `email_verified` claim, and users who
> attempt to sign in without a verified email address will be prevented from doing so:

Reviewer 2: ```suggestion
By default, Teleport validates the `email_verified` claim, and users who
attempt to sign in without a verified email address are prevented from doing so:
```

```text
ERROR: SSO flow failed.
identity provider callback failed with error: OIDC provider did not verify email.
        email not verified by OIDC provider
```

> For testing and other purposes, you may opt out of this behavior by enabling

Reviewer 2: ```suggestion
For testing and other purposes, you can opt out of this behavior by enabling
```
